### ISSUE TITLE
Support mapping for Discriminator object.

### ISSUE THREAD

**Issue by kw7oe** [juhaku/utoipa#324]
Hi, first of all thanks for the amazing work 💯!

I would like to work on supporting defining custom mapping for Discriminator object. Here's a [minimally working PR](https://github.com/getditto/utoipa/pull/1/files) on a fork according to the below proposal (**TLDR;** `#[schema(mapping = "...")]`). 

The PR only works for Unnamed Field Struct and hardcoding the value of the mapping. If it's keen to have this feature added as part of utoipa, do let me know and I'm happy to open a PR here and work on it according to any suggestions.

### Context 
Here's a example use case:

```rust
#[derive(ToSchema, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateResult {
    pub id: String,
    pub amount: u64
}

#[derive(ToSchema, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UpdateResult {
    pub id: String,
    pub amount: u64
}

#[derive(ToSchema, Debug, Serialize, Deserialize)]
#[serde(tag = "method", rename_all = "camelCase")]
pub enum Result {
    Create(CreateResult),
    Update(UpdateResult),
}
```

Using `utoipa`, the following would be generated:

```json
{
  "oneOf": [
    {
      "allOf": [
        {
          "$ref": "#/components/schemas/CreateResult"
        },
        {
          "type": "object",
          "required": [
            "method"
          ],
          "properties": {
            "method": {
              "type": "string",
              "enum": [
                "create"
              ]
            }
          }
        }
      ]
    },
    {
      "allOf": [
        {
          "$ref": "#/components/schemas/DeleteResult"
        },
        {
          "type": "object",
          "required": [
            "method"
          ],
          "properties": {
            "method": {
              "type": "string",
              "enum": [
                "delete"
              ]
            }
          }
        }
      ]
    }
  ],
  "discriminator": {
    "propertyName": "method"
  }
}
```

and the expected json will be as follow:

```json
{
  "method": "create | delete",
  "id": "1",
}
```

However. according to the [Discriminator spec](https://spec.openapis.org/oas/latest.html#discriminator-object), it would need to method would need to match either `CreateResult` or `DeleteResult`. This is where `mapping` come in and help with it. Without the custom mapping, the generated client using OpenAPI geneartor will not work as expected.

### Proposal

Support additonal `mapping` attribute in `schema` macro:

```rust
#[derive(ToSchema, Debug, Serialize, Deserialize)]
#[serde(tag = "method", rename_all = "camelCase")]
pub enum Result {
    #[schema(mapping = "create")]
    Create(CreateResult),
    #[schema(mapping = "update")]
    Update(UpdateResult),
}
```
which will generate the following:

```json
{
  "oneOf": [
    {
      "allOf": [
        {
          "$ref": "#/components/schemas/CreateResult"
        },
        {
          "type": "object",
          "required": [
            "method"
          ],
          "properties": {
            "method": {
              "type": "string",
              "enum": [
                "create"
              ]
            }
          }
        }
      ]
    },
    {
      "allOf": [
        {
          "$ref": "#/components/schemas/DeleteResult"
        },
        {
          "type": "object",
          "required": [
            "method"
          ],
          "properties": {
            "method": {
              "type": "string",
              "enum": [
                "delete"
              ]
            }
          }
        }
      ]
    }
  ],
  "discriminator": {
    "propertyName": "method",
    "mapping": {
      "create": "#/components/schemas/CreateResult",
      "delete": "#/components/schemas/DeleteResult"
    }
  }
}
```
I'm not sure if this is the best way so do let me know if there's any better alternatives. Some of the inputs I might need is how do we handle the different kind of Enums we have and also what if those reference struct are not derive with `ToSchema`.

**Comment by juhaku** [comment]
Hello, anytime it's my pleasure to work on this crate. :slightly_smiling_face: 

Yes sure all contributions are welcome. :+1: 

Took a look at the specs and found a line stating that _for inline schemas the discriminator field will not be considered_.  This has two implications where __one__ is that the current discriminator field implementaion is incomplete and __two__ there is no need to consider other type's of enums. 
I think since we are currently creating the `discriminator` field only when there is serde's `tag` attribute over the enum we can directly use the variant value as field to the mapping if variant is tuple style with single argument without having need for custom `#mapping(...)` attribute. 
```rust
#[derive(ToSchema, Debug, Serialize, Deserialize)]
#[serde(tag = "method", rename_all = "camelCase")]
pub enum Result {
    Create(CreateResult),
    Update(UpdateResult),
}
```
We could directly use the `CreateResult` and`UpdateResult` in generation of the `mapping`.

In code we have access to the variant value so we can create a ref directly from it with something like this. (not fully working code)
```rust
let name = &*variant.ident.to_string();
Ref::from_schema_name(name);
```

If we had enum like one below, we could just ingore the mapping for fields which does not comply with the spec of being tuple with single argument reference.
```rust
enum Foobar {
   UnitValue,
   StructValue {
     field: String
   },
   Tuple(ValueObject),
}
```
If the `ValueObject` would not be a `ToSchema` component users should decoreate it with _value_type` attribute e.g. `#[schema(value_type = String)]`. However this would result invalid spec same as not defining instead without ugly error in the OpenAPI doc that _Component #/components/schema/NotASchema not found_. Because I reckon primitive type object would not work correctly as a mapping reference. Nevertheless one could use the `value_type` attribute to define another `ToSchema` compoent instead of the value which is not a `ToSchema` component.

**Comment by jamievaughan** [comment]
Has there been any movement on this? I'm having issues generating clients from the swagger output which expects a mapping for polymorphic types. If not, I'm happy to pick this up!

**Comment by kw7oe** [comment]
> Has there been any movement on this? I'm having issues generating clients from the swagger output which expects a mapping for polymorphic types. If not, I'm happy to pick this up!

I haven't have the time to contribute back to upstream since the suggested approach is slight different from what I initially implemented. My implementation is available over [here](https://github.com/getditto/utoipa/pull/1/files) but it's quite far behind the latest branch.

So feel free to take on this!

**Comment by juhaku** [comment]
@jamievaughan @kw7oe  I commented on another issue related to this discriminator thing. I'd like some ideas of regarding how to solve this, see https://github.com/juhaku/utoipa/issues/621#issuecomment-2359369248. Along the way if the discriminator field will be solved the mappings could be experimented as well.

**Comment by juhaku** [comment]
The new way of handling discriminator is implemented in #1059 and it supports custom mapping. However it does not automatically collect mapping items and they need to be manually provided if wanted. Yet this is planned and probably will be implemented in future PRs.

**Comment by juhaku** [comment]
Link to taks: https://github.com/users/juhaku/projects/1?pane=issue&itemId=80603415
